### ISSUE TITLE
Switch Zone: Create "late-binding" mechanism for provisioning underlay address

### ISSUE THREAD

**Issue by smklein** [oxidecomputer/omicron#1961]
See: https://github.com/oxidecomputer/omicron/issues/1960 for context

- We want to launch the switch zone before underlay addresses are established
- When the underlay comes up, we'd like to add the "switch zone address" to the already-running "switch zone"

**Comment by smklein** [comment]
@Nieuwejaar I wanted to get your feedback on this, here's my proposal:

1. The bootstrap service launches the switch zone *without* an underlay IP address. Services in the switch zone will need to be updated to ensure that this address is an optional argument.
2. The underlay is established as a part of RSS, eventually sled agents can boot.
3. Once the sled agent starts, it **halts and destroys** the switch zone, and then **creates a new switch zone** using the underlay address.

WDYT? This is definitely doable from the sled-agent side of things, and seems like the easiest path forward (it seems doable in https://github.com/oxidecomputer/omicron/pull/1970 ). Does stopping / re-starting the switch zone pose any problems from your point-of-view (aside from https://www.illumos.org/issues/15167 , perhaps)?

**Comment by Nieuwejaar** [comment]
For dpd and tfportd, the address is already optional.  If one isn't provided, they use localhost and the default port number.

Tearing down and restarting the switch zone should be fine - modulo the illumos kernel panic.  If you can `svcadm disable` tfport before halting the zone, I can have it clean up all of its links before exiting.  If all goes well, that _should_ avoid the kernel panic.

**Comment by bnaecker** [comment]
@Nieuwejaar I'm surprised that this is workable. Is there really _no_ state at all maintained by the programs running in the switch zone? It seems like restarting either `dpd` or the BF SDE would walk back through a bunch of link initialization that might interrupt things. Is that not true?

As an alternative, we could include in the `dpd` Dropshot API an endpoint for starting a _new_ server listening on some underlay address. Those would be duplicates really, modulo that endpoint for starting a new server, each providing the same API just on different addresses. IIRC, the inner state of the main `dpd` server object (or the `Switch` type maybe?) is behind a lock already, so I don't think there should be an additional hit for synchronizing requests handled by different servers.

**Comment by Nieuwejaar** [comment]
Sorry - I didn't mean to imply that this would be seamless.  The switch would essentially go offline during the restart and the links would all need to come back up.  By "fine" I just meant that this would work, not that it was ideal.  Thanks for catching me.

We should be able to support multiple servers.  The code is all written assuming that dropshot can have multiple requests in flight at a time.  Whether that is due to multiple single-threaded servers or one multi-threaded server shouldn't matter.

**Comment by bnaecker** [comment]
Oh, ok, that's good to know! I definitely just misinterpreted the strength of what you were saying then. Thanks for clarifying!

**Comment by rmustacc** [comment]
I would really try to avoid destroying and restarting this, causing everything to have a link flap and cause a lot of chaos. In general, I would instead just look at things like an SMF refresh method to do this and cause it to rebind or do something else.

In particular, the switch would still be programmed to do a bunch of stuff and would keep doing it even with the zone gone, unless we turned everything off explicitly and shut down the ports. So while shutting down and recreating is expedient, I'm not sure it's what we want.

**Comment by smklein** [comment]
So I filed https://github.com/oxidecomputer/omicron/issues/1898 a while back, and I think this discussion of "should we restart the whole zone" vs "should we restart the services within that zone with a new config" is related. I wanted to bring that up to explain rationale, but ultimately, availability wins, so I'm happy to prioritize whatever helps us meet that goal.

While working to resolve #1898 , we discussed injecting a profile into `/var/svc/profile/site.xml` *before* a zone is started, so SMF services could read their configuration without the sled agent needing to `zlogin` and issue requests. It was my understanding that, as a principle, we should avoid modifying the filesystems of non-global zones while they're running.

This was a big reason I wanted to "kill and restart" the switch zone -- it would give the sled agent an opportunity to inject this profile without the switch zone seeing anything. 

We definitely could create a new profile, including the underlay address for the switch zone as that's available, and use `rename` to atomically update `/var/svc/profile/site.xml` while the zone is running. Individually refreshing the SMF services wouldn't be atomic, but maybe that's fine.

**Comment by rmustacc** [comment]
I think there are a lot of different ways to achieve this without actually injecting something into the zone that way. They do involve a sled agent reaching in, but they don't require you to actually adjust a profile and that isn't the way I would try to communicate that in general in the case of a running system.

That is, I believe we can keep the file system immutable in that regard and still one can inject run-time information. Would it help to talk through a couple different strategies for this? It might help me better understand why the profile route is the preferred path for this case. Just as an example, there's nothing that stops us from just setting the requisite properties on the SMF instance itself and then calling a refresh method.

Ultimately, if the goal is to never reach into the zone via sled agent at all, then I can get that. I think we just then have to weigh what that means about the switch itself and critically if we're going to shut off the zone, we also probably want to shut off the ASIC or at least unconfigure that if it's the way folks would prefer to go.

**Comment by smklein** [comment]
> That is, I believe we can keep the file system immutable in that regard and still one can inject run-time information. Would it help to talk through a couple different strategies for this? It might help me better understand why the profile route is the preferred path for this case. Just as an example, there's nothing that stops us from just setting the requisite properties on the SMF instance itself and then calling a refresh method.

I went ahead and implemented this in https://github.com/oxidecomputer/omicron/pull/1970/commits/043d580e45aa457069d56fcdfa101caed484d5f8 .

**Comment by smklein** [comment]
Now that #1970 has been merged, I'm still leaving this open -- we are leaving the zone up, but restarting services to provide the new underlay address value.

The "SMF refresh" method by default does nothing. Looking around, having our services interpret "SIGHUP" as a sign to reload configuration seems common, so we could implement signal handler that in our services that want to provide an option for being reloaded?

**Comment by rmustacc** [comment]
> Now that #1970 has been merged, I'm still leaving this open -- we are leaving the zone up, but restarting services to provide the new underlay address value.
> 
> The "SMF refresh" method by default does nothing. Looking around, having our services interpret "SIGHUP" as a sign to reload configuration seems common, so we could implement signal handler that in our services that want to provide an option for being reloaded?

Yeah, the default refresh behavior is not specified. There are lots of ways we could do this. It could be a signal (I think you probably want a `SIGUSR1` perhaps over `SIGHUP`), but you could also hit an API endpoint that causes it to reread its service properties or anything else you want. A signal is simple, but it certainly doesn't have to be that.
